### PR TITLE
docs: add a next-phase CRD design document

### PR DESCRIPTION
**Pull request by phlogistonjohn**
I've been working on this in the background and I think it's good enough to share now. I do not expect to merge this as-is but to mainly use it to kick off additional discussion. That said, I think it would be good to merge it once we agree on the broad strokes and then continue to modify it there.

Please, do ask questions and make me justify some of these ideas! :-)

Part of: #2

### CODE REVIEW

File: docs/design/crd-proposal-phase1.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+* SmbSecurityDomain - A CR that encapsulates the knowledge needed to create "local" users or bind to an Active Directory domain
+
+The operator will take the SmbShare and SmbSecurityDomain resources as inputs
+and create as many smbd, winbind, etc. backing services as needed. Users
+will have limited input into what backing services the operator will create.
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
+SmbSecurityDomain refernce will define the main security properties of the
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:28]: `refernce` ==> `reference`

File: docs/design/crd-proposal-phase1.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  * `backend` - enumerated string - the name of a (supported) ID mapping backend
+  * `domainName` - string - the name of the domain being configured
+  * TBD - options to configure the backend, ID range, etc.
+
+
+## SmbShare
+
+Spec Options:
+* `shareName` - string - Optional string giving the share a full SMB compliant name. If unset, the name will be derived by the operator.
+* `storage` - mapping - A new subsection to configure the storage backing the share.  * `pvc` - mapping - A new subsection used to configure PVC backing storage.
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:57]: newline missing

File: docs/design/crd-proposal-phase1.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    * `name` - string - The name of a PVC, managed externally to the operator, for the share
+    * `spec` - mapping, embedded pvc spec - An embedded PVC spec that will be used to dynamically create a backing PVC for the share; sharing the lifecycle of the PVC with the share
+  * TBD - Any other more custom storage backends if needed
+* `securityDomain`  - string - The name of the SmbSecurityDomain CR associated with this share
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:62]: What do you imagine the value of `grouping`to look like? A list of other smbshare names?

File: docs/design/crd-proposal-phase1.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+* `clustering` - mapping - A new subsection pertaining to clustering smbd instances
+  * `mode` - enumerated string "none" or "ctdb" - This option tells the operator if this share should be clustered or not
+  * `minSize` - int - Minimum number of smbd instances when clustering
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:65]: Why minimum and not an exact size?

File: docs/design/crd-proposal-phase1.md (hunk 3 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:62]: OK, I see it in the examples below. Just a group name that will have all the shares with the same grouping name combined under on smbd instance.

File: docs/design/crd-proposal-phase1.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+Spec Options:
+* `shareName` - string - Optional string giving the share a full SMB compliant name. If unset, the name will be derived by the operator.
+* `storage` - mapping - A new subsection to configure the storage backing the share.  * `pvc` - mapping - A new subsection used to configure PVC backing storage.
+    * `name` - string - The name of a PVC, managed externally to the operator, for the share
+    * `spec` - mapping, embedded pvc spec - An embedded PVC spec that will be used to dynamically create a backing PVC for the share; sharing the lifecycle of the PVC with the share
+  * TBD - Any other more custom storage backends if needed
+* `securityDomain`  - string - The name of the SmbSecurityDomain CR associated with this share
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
+* `clustering` - mapping - A new subsection pertaining to clustering smbd instances
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:63]: We might want to structure this a little differently, not yet sure how: As mentioned I was identifying 3 aspects of samba's dbwrap backend currently provided by ctdb
1. indication of cluster intactness in the backend (recovery lock file)
2. volatile DBs (and messaging)
3. persistent DBs.

of these, #1 might be useful without clustering, #3 is definitely useful for making config and settings (join...) persistent. But #2 would really for moving to all-active clustering. This would require development in samba though. 

So to summarize, not sure if putting this under "clustering" is entirely sufficient or appropriate.

**Review by obnoxxx — COMMENTED** [review]
In general, I like the idea of structuring it this way very much. A couple of comments and thoughts inside.

File: docs/design/crd-proposal-phase1.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
+* `localName` - string - Optional string used to indicate how this instance will
+  be referred to in an Active Directory
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
+  * `backend` - enumerated string - the name of a (supported) ID mapping backend
+  * `domainName` - string - the name of the domain being configured
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:49]: The `realm` translates to the smb.conf parameter of the same name. The `domainName` keys loosely map to the domain key of the `idmap` smb.conf settings. These are optional but allow one to say configure a non-default idmap setting like "ad" for either the primary `realm` or other trusted domains.

File: docs/design/crd-proposal-phase1.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+## SmbSecurityDomain
+
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
+* `localName` - string - Optional string used to indicate how this instance will
+  be referred to in an Active Directory
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:47]: It is supposed to contain as many settings for domains as you have (primary or trusted) that you want to configure.

Example:
```
domains:
  - domainName: myrealm.mycorp.example.com
    backend: ad
  - domainName: atrust.mycorp.example.com
    backend: ad
```

Do you you think it would be clearer as a map-of-maps like this example:
```
domains:
    myrealm.mycorp.example.com:
        backend: ad
    atrust.mycorp.example.com:
        backend: ad
```?

If we need to the section could include other parameters such as idmap ranges, etc. But I left it out of the initial draft becase (a) I was hoping to keep this siimple and (b) hoping the operator could do the heavy lifting of determining valid ranges after being simply told to use the ad (and check for rfc2307 in the ad).

File: docs/design/crd-proposal-phase1.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+* SmbShare - A CR that encapsulates all the information needed to export a single share
+* SmbSecurityDomain - A CR that encapsulates the knowledge needed to create "local" users or bind to an Active Directory domain
+
+The operator will take the SmbShare and SmbSecurityDomain resources as inputs
+and create as many smbd, winbind, etc. backing services as needed. Users
+will have limited input into what backing services the operator will create.
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:27]: It would not. This is poorly worded. I will improve it.

File: docs/design/crd-proposal-phase1.md (hunk 4 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:65]: I wasn't sure if we'd want to have a minSize and a maxSize in the future. But perhaps that is premature. Would you prefer if I changed it to just "size" and we can deal with the future in the future?

File: docs/design/crd-proposal-phase1.md (hunk 5 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:63]: Sure. At the time I was only thinking of non-clustered or clustered. I do think that some sort of subsection would be needed to what kind of deployment (in the general sense, not just the k8s resource) we want and how it is expected to scale. Perhaps if we replace "clustering" with "scaling" as a more generic term? And then the type of scaling might be single instance or clustered or some hybrid...

Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:66]: Also, I forgot to note. None of this is discussing what the defaults would be unless explicitly stated. I suspect that in some cases we might prefer to have the more "complex" mode the default but allow some use cases to "opt-out". For example if a clustered mode was the default, a user might want to scale down to one instance for a read only share of limited importance to save on resources, etc.

File: docs/design/crd-proposal-phase1.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md]: updated.

File: docs/design/crd-proposal-phase1.md (hunk 2 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md]: fixed.

File: docs/design/crd-proposal-phase1.md (hunk 1 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md]: should be fixed.

**Comment by phlogistonjohn** [comment]
Pushed an updated version that fixes some of the items that didn't warrant additional discussion. I'm looking forward to hearing your opinions on the remaining items!

File: docs/design/crd-proposal-phase1.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:49]: I think a tag for the name is fine, but `domainName` under `domains` is kind of redundant. we could just have `name` under `domains`?

File: docs/design/crd-proposal-phase1.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:51]: Got it.

**Review by obnoxxx — COMMENTED** [review]
After your comments and thinking about this a bit more, I think we should rename `SmbSecurityDomain`: This fundamentally corresponds to the global configuration of a samba server (`smb.conf` file). The `SmbShare` type corresponds to a share section in an `smb.conf` file. 

So in this regard, what you named `SmbSecurityDomain` might better be named `SambaInstance` or `SambaServer`. The instance is ultimately identified by its internal ID (the machine SID) once the server/db is initialised. The fqdn (essentially `localName` + `realm`) can also identify the instance.

There can only be one main smbd for a given SID / fqdn. So the `SmbShare` CRs need to end up as share config sections for their respective smbd instances.

Each `SmbShare` needs to be associated to one and only one `SambaServer` (or `SambaSecurityDomain`). Also very important that (1) each PVC is only attached to only one `SambaServer` (indirectly). If there is a possibility for a subdirectory of a PVC to be turned into another PVC, then it is important that these two PVCs are attached to the *same* `SambaServer`.

File: docs/design/crd-proposal-phase1.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:49]: Sure. I thought a more descriptive key would be better but it is somewhat redundant.

**Comment by phlogistonjohn** [comment]
> After your comments and thinking about this a bit more, I think we should rename `SmbSecurityDomain`: This fundamentally corresponds to the global configuration of a samba server (`smb.conf` file). The `SmbShare` type corresponds to a share section in an `smb.conf` file.
> 

I agree that the security & domain settings reside in the global section of the smb.conf. I also agree that SmbShare aligns with the share section of smb.conf.


> So in this regard, what you named `SmbSecurityDomain` might better be named `SambaInstance` or `SambaServer`. The instance is ultimately identified by its internal ID (the machine SID) once the server/db is initialised. The fqdn (essentially `localName` + `realm`) can also identify the instance.
> 

I was trying to explicitly avoid tying the operator's config to either server instances or smb.conf. The names of the many config items reflect the parameters in smb.conf my intention is not to pass these options to smb.conf in a  1:1 manner. Rather the names are used for familiarity.  

I prefer SmbSecurityDomain because:
* It only covers the security and domain settings, while real smb.conf has many many more global options
* It emphasizes the important user provided nature of security settings
* It manages some important security and domain settings that typically lie outside of smb.conf - mainly the reference to a secret holding user information for non-domain-joined mode


> There can only be one main smbd for a given SID / fqdn. So the `SmbShare` CRs need to end up as share config sections for their respective smbd instances.
> 

Yes.

> Each `SmbShare` needs to be associated to one and only one `SambaServer` (or `SambaSecurityDomain`). 

Agree, when talking about a non-clustered scenario. For clustering its a tad bit ambiguous how to describe this as I assume you see the mutliple clustered smbd as sharing one config, right?

> Also very important that (1) each PVC is only attached to only one `SambaServer` (indirectly). If there is a possibility for a subdirectory of a PVC to be turned into another PVC, then it is important that these two PVCs are attached to the _same_ `SambaServer`.

Again, I'm a little unclear on how you think about this in a clustered situation. I envision one Read-Write-Many PVC would be attached to as many pods as needed. I agree that this PVC would be coupled to one (SmbShare, SmbSecurityDoamin) instance, however many smbds/pods that may entail.

File: docs/design/crd-proposal-phase1.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:49]: changed.

File: docs/design/crd-proposal-phase1.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+create shares that can be consumed by a variety of clients for various use
+cases. Some sections are deliberately marked "TBD" (to be decided) as
+placeholders for more detail that will be needed even in Phase 1.
+
+## SmbSecurityDomain
+
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
```
Comment (gd) [inline docs/design/crd-proposal-phase1.md:45]: What exactly does the 'secret' string refer to? Is it some sort of embedding user and group definitions in a  cryptographical manner?

File: docs/design/crd-proposal-phase1.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+```yaml
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbSecurityDomain
+metadata:
+  name: "local-users"
+spec:
+  mode: "user"
+  users:
+    secret: "local-users-secret"
```
Comment (gd) [inline docs/design/crd-proposal-phase1.md:88]: Just curious: What is that secret used for?

File: docs/design/crd-proposal-phase1.md (hunk 9 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:45]: In kubernets a secret is a bundle of configuration that may (or may not) be protected by encyption. By default it is simply obscured but you can install add ons that can protect the secret contents more thoroughly.

File: docs/design/crd-proposal-phase1.md (hunk 10 above)
Comment (phlogistonjohn) [reply docs/design/crd-proposal-phase1.md:88]: I envision it would contain a json configuration of local users and groups, the actual structure of said json TBD of course. :-)

**Comment by phlogistonjohn** [comment]
After a verbal discussion I came to the realization that 'localName" was not appropriate for a CR that would be usable by multiple instances. I have renamed the field "instanceNamePrefix" and updated the description.

**Comment by phlogistonjohn** [comment]
As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.

File: docs/design/crd-proposal-phase1.md (hunk 4 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:67]: Note from earlier discussion: minSize was chosen to allow for more flexible grouping...

**Comment by obnoxxx** [comment]
> As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.

I think that `SmbIdentity` would really mean that we'd have a 1:1 correspondence between this and an instance. An "Identity" could conceptually by its name not serve as a template for instances.

Maybe call it `SambaSecurityConfig` or `SambaSecuritySetup` instead?

@phlogistonjohn

**Comment by phlogistonjohn** [comment]
> > As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.
> 
> I think that `SmbIdentity` would really mean that we'd have a 1:1 correspondence between this and an instance. An "Identity" could conceptually by its name not serve as a template for instances.
> 
> Maybe call it `SambaSecurityConfig` or `SambaSecuritySetup` instead?
> 

I'd go with SmbSecurityConfig to match the other CR name. But I'm fine with ...Config as the suffix rather than "Domain".

**Comment by obnoxxx** [comment]
> I'd go with `SmbSecurityConfig` to match the other CR name. But I'm fine with ...Config as the suffix rather than "Domain".

Oh, sure, I mistyped Samba for Smb, `SmbSecurityConfig` is good!

**Comment by phlogistonjohn** [comment]
I have pushed a new version. 
It changes SmbSecurityDomain to SmbSecurityConfig.
All options pertaining to clustering have been moved under "scaling" and the names are tweaked but  are functionally the same.
Grammar and spelling improvements.
Block formatting on some of the long lines.

File: docs/design/crd-proposal-phase1.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+  storage:
+    pvc:
+      name: "docs"
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "local-users"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:135]: `securityConfig`

File: docs/design/crd-proposal-phase1.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+      name: "cool-ad.int.example.com"
+    - backend: ad
+      name: "my-trust.int.example.com"
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "cool-domain"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:166]: `securityConfig`

File: docs/design/crd-proposal-phase1.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+  storage:
+    pvc:
+      name: "docs"
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "cool-domain"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:176]: `securityConfig`

File: docs/design/crd-proposal-phase1.md [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+Two shares, with groups:
+```yaml
+# SmbSecurityConfig not shown
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "local-users"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:194]: `securityConfig`

File: docs/design/crd-proposal-phase1.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+  storage:
+    pvc:
+      name: "docs"
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "local-users"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:207]: `securityConfig`

File: docs/design/crd-proposal-phase1.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+One share with clustering enabled:
+```yaml
+# SmbSecurityConfig not shown
+---
+apiVersion: samba-operator.samba.org/v1alpha1
+kind: SmbShare
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "cool-domain"
```
Comment (obnoxxx) [inline docs/design/crd-proposal-phase1.md:228]: `securityConfig`

File: docs/design/crd-proposal-phase1.md (hunk 11 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:135]:
```suggestion
  securityConfig: "local-users"
```

File: docs/design/crd-proposal-phase1.md (hunk 12 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:166]:
```suggestion
  securityConfig: "cool-domain"
```

File: docs/design/crd-proposal-phase1.md (hunk 13 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:176]:
```suggestion
  securityConfig: "cool-domain"
```

File: docs/design/crd-proposal-phase1.md (hunk 14 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:194]:
```suggestion
  securityConfig: "local-users"
```

File: docs/design/crd-proposal-phase1.md (hunk 15 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:207]:
```suggestion
  securityConfig: "local-users"
```

File: docs/design/crd-proposal-phase1.md (hunk 16 above)
Comment (obnoxxx) [reply docs/design/crd-proposal-phase1.md:228]:
```suggestion
  securityConfig: "cool-domain"
```

**Review by obnoxxx — CHANGES_REQUESTED** [review]
Looks good in general.
A few cases in the examples with the old names.

**Review by obnoxxx — COMMENTED** [review]
Oh, and is the duplicate signoff in the commit message intentional?

**Comment by phlogistonjohn** [comment]
> Oh, and is the duplicate signoff in the commit message intentional?

Heh. No. It's a side effect of something I am experimenting with.

**Review by obnoxxx — APPROVED** [review]
lgtm
